Helper/PCAHelper.cs

```diff
@@ -0,0 +1,220 @@
+        /// Interactive attempt is optional.
+        /// It provides optional delegates to customize behavior.
+        /// </summary>
+        /// <param name="doSilent">Determines whether to execute AcquireTokenSilent</param>
+        /// <param name="doInteractive">Determines whether to execute AcquireTokenInteractive. By detault, UI interaction takes place if silent action fails.</param>
+        /// <param name="preferredAccount">Function that determines the account to be used. The default is first. (optional)</param>
+        /// <param name="customizeSilent">This is a delegate to optionally customize AcquireTokenSilentParameterBuilder.</param>
+        /// <param name="customizeInteractive">This is a delegate to optionally customize AcquireTokenInteractiveParameterBuilder.</param>
+        /// <returns>Authenitcation result</returns>
+        public async Task<AuthenticationResult> EnsureAuthenticatedAsync(
+                                                                bool doSilent = true,
+                                                                bool doInteractive = true,
+                                                                Func<IEnumerable<IAccount>, IAccount> preferredAccount = null,
+                                                                Action<AcquireTokenSilentParameterBuilder> customizeSilent = null,
+                                                                Action<AcquireTokenInteractiveParameterBuilder> customizeInteractive = null)
+        {
+            AuthResult = null;
+
+            try
+            {
+                if (doSilent)
+                {
+                    IAccount account = null;
+                    IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync().ConfigureAwait(false);
+                    if (preferredAccount != null)
+                    {
+                        account = preferredAccount(accounts);
+                    }
+                    else if(accounts != null)
+                    {
+                        account = accounts.FirstOrDefault();
+                    }
+
+                    // Customize silentBuilder
+                    var silentparamsBuilder = PCA.AcquireTokenSilent(_scopes, account);
+                    if (customizeSilent != null)
+                    {
+                        customizeSilent(silentparamsBuilder);
+                    }
+
+                    AuthResult = await silentparamsBuilder.ExecuteAsync()
+                                          .ConfigureAwait(false);
+                }
+                else if (doInteractive)
+                {
+                    await AcquireInteractive(customizeInteractive).ConfigureAwait(false);
+                }
+                else
+                {
+                    throw new ArgumentException($"Both doSilent and do Interactive cannot be false");

```

Comment:
nit: I don't think these should need to be set, we should just do them automatically.

Reply:
I am not understanding here. What does not need to be set?

Reply:
Your helper should decide on its own if interactive auth is needed or not. It should not be the responsibility of the user.

Reply:
OK.